svm/src/account_loader.rs

```diff
@@ -2105,4 +2131,97 @@ mod tests {
         assert_eq!(account.rent_epoch(), 0);
         assert_eq!(account.lamports(), 0);
     }
+
+    // Ensure `TransactionProcessingCallback::inspect_account()` is called when
+    // loading accounts for transaction processing.
+    #[test]
+    fn test_inspect_account_non_fee_payer() {
+        let mut mock_bank = TestCallbacks::default();
+
+        let address0 = Pubkey::new_unique(); // <-- fee payer
+        let address1 = Pubkey::new_unique(); // <-- initially alive
+        let address2 = Pubkey::new_unique(); // <-- initially dead
+        let address3 = Pubkey::new_unique(); // <-- program
+
+        let mut account0 = AccountSharedData::default();
+        account0.set_lamports(1_000_000_000);
+        mock_bank.accounts_map.insert(address0, account0.clone());
+
+        let mut account1 = AccountSharedData::default();
+        account1.set_lamports(2_000_000_000);
+        mock_bank.accounts_map.insert(address1, account1.clone());
+
+        // account2 *not* added to the bank's accounts_map
+
+        let mut account3 = AccountSharedData::default();
+        account3.set_lamports(4_000_000_000);
+        account3.set_executable(true);
+        account3.set_owner(native_loader::id());
+        mock_bank.accounts_map.insert(address3, account3.clone());
+
+        let message = Message {
+            account_keys: vec![address0, address1, address2, address3],
+            header: MessageHeader::default(),
```

Comment:
So I guess this default header is causing all accounts to be loaded as writable because it means there are 0 signers and 0 readonly non-signers? And the program is loaded as readonly because it was demoted by `demote_program_id`? Do you mind adding comments describing this behavior?

Reply:
Because it's important that account inspection correctly captures whether an account is loaded with a write lock, right?

Reply:
Ok that seems reasonable to me